File: backend/src/main/java/com/leadgrowth/controller/ReportController.java
package com.leadgrowth.controller;

import com.leadgrowth.dto.LeadDto;
import com.leadgrowth.entity.Campaign;
import com.leadgrowth.service.CampaignService;
import com.leadgrowth.service.ExportService;
import com.leadgrowth.service.LeadService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.leadgrowth.dto.DailyReportSubmitRequest;
import com.leadgrowth.dto.ReportDto;
import com.leadgrowth.dto.ReportReviewRequest;
import com.leadgrowth.service.ReportService;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/reports")
public class ReportController {

    private final ExportService exportService;
    private final CampaignService campaignService;
    private final LeadService leadService;
    private final ReportService reportService;

    public ReportController(ExportService exportService, CampaignService campaignService, LeadService leadService, ReportService reportService) {
        this.exportService = exportService;
        this.campaignService = campaignService;
        this.leadService = leadService;
        this.reportService = reportService;
    }

    @PostMapping("/daily")
    public ResponseEntity<ReportDto> submitDailyReport(@RequestBody DailyReportSubmitRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(reportService.submitDailyReport(request, email));
    }

    @GetMapping("/my-reports")
    public ResponseEntity<List<ReportDto>> getMyReports() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(reportService.getUserReports(email));
    }

    @GetMapping("/workspace-reports")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<List<ReportDto>> getWorkspaceReports() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(reportService.getPendingWorkspaceReports(email));
    }

    @PatchMapping("/{id}/review")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<ReportDto> reviewReport(
            @PathVariable Long id,
            @Valid @RequestBody ReportReviewRequest request
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(reportService.reviewReport(id, request, email));
    }

    // --- Campaign Report Endpoints ---

    @GetMapping("/campaigns/csv")
    public ResponseEntity<byte[]> downloadCampaignsCsv() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        List<Campaign> campaigns = campaignService.getCampaigns(email);
        byte[] data = exportService.exportCampaignsToCsv(campaigns);
        
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=campaigns.csv")
                .contentType(MediaType.parseMediaType("text/csv"))
                .body(data);
    }

    @GetMapping("/campaigns/excel")
    public ResponseEntity<byte[]> downloadCampaignsExcel() throws Exception {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        List<Campaign> campaigns = campaignService.getCampaigns(email);
        byte[] data = exportService.exportCampaignsToExcel(campaigns);
        
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=campaigns.xlsx")
                .contentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"))
                .body(data);
    }

    @GetMapping("/campaigns/pdf")
    public ResponseEntity<byte[]> downloadCampaignsPdf() throws Exception {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        List<Campaign> campaigns = campaignService.getCampaigns(email);
        byte[] data = exportService.exportCampaignsToPdf(campaigns);
        
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=campaigns.pdf")
                .contentType(MediaType.APPLICATION_PDF)
                .body(data);
    }

    // --- Lead Report Endpoints ---

    @GetMapping("/leads/csv")
    public ResponseEntity<byte[]> downloadLeadsCsv() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        List<LeadDto> leads = leadService.getLeads(email);
        byte[] data = exportService.exportLeadsToCsv(leads);
        
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=leads.csv")
                .contentType(MediaType.parseMediaType("text/csv"))
                .body(data);
    }

    @GetMapping("/leads/excel")
    public ResponseEntity<byte[]> downloadLeadsExcel() throws Exception {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        List<LeadDto> leads = leadService.getLeads(email);
        byte[] data = exportService.exportLeadsToExcel(leads);
        
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=leads.xlsx")
                .contentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"))
                .body(data);
    }

    @GetMapping("/leads/pdf")
    public ResponseEntity<byte[]> downloadLeadsPdf() throws Exception {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        List<LeadDto> leads = leadService.getLeads(email);
        byte[] data = exportService.exportLeadsToPdf(leads);
        
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=leads.pdf")
                .contentType(MediaType.APPLICATION_PDF)
                .body(data);
    }

    @GetMapping("/leads/{id}/pdf")
    public ResponseEntity<byte[]> downloadSingleLeadPdf(@PathVariable Long id) {
        byte[] data = exportService.exportSingleLeadPdf(id);
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=lead_" + id + "_dossier.pdf")
                .contentType(MediaType.APPLICATION_PDF)
                .body(data);
    }
}

File: backend-dotnet/Controllers/ReportController.cs
using System.Security.Claims;
using LeadGrowth.DTOs;
using LeadGrowth.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace LeadGrowth.Controllers;

[ApiController]
[Route("api/reports")]
[Authorize]
public class ReportController : ControllerBase
{
    private readonly IReportService _reportService;
    private readonly IExportService _exportService;
    private readonly ICampaignService _campaignService;
    private readonly ILeadService _leadService;
    private readonly IWorkspaceService _workspaceService;

    public ReportController(
        IReportService reportService,
        IExportService exportService,
        ICampaignService campaignService,
        ILeadService leadService,
        IWorkspaceService workspaceService)
    {
        _reportService = reportService;
        _exportService = exportService;
        _campaignService = campaignService;
        _leadService = leadService;
        _workspaceService = workspaceService;
    }

    [HttpPost("daily")]
    public async Task<ActionResult<ReportDto>> SubmitDailyReport([FromBody] DailyReportSubmitRequest request)
    {
        var email = GetUserEmail();
        try
        {
            var report = await _reportService.SubmitDailyReportAsync(request, email);
            return Ok(report);
        }
        catch (KeyNotFoundException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("my-reports")]
    public async Task<ActionResult<List<ReportDto>>> GetMyReports()
    {
        var email = GetUserEmail();
        var reports = await _reportService.GetUserReportsAsync(email);
        return Ok(reports);
    }

    [HttpGet("workspace-reports")]
    [Authorize(Policy = "RequireManagerOrAdmin")]
    public async Task<ActionResult<List<ReportDto>>> GetWorkspaceReports()
    {
        var email = GetUserEmail();
        var reports = await _reportService.GetPendingWorkspaceReportsAsync(email);
        return Ok(reports);
    }

    [HttpPatch("{id}/review")]
    [Authorize(Policy = "RequireManagerOrAdmin")]
    public async Task<ActionResult<ReportDto>> ReviewReport(long id, [FromBody] ReportReviewRequest request)
    {
        var email = GetUserEmail();
        try
        {
            var report = await _reportService.ReviewReportAsync(id, request, email);
            return Ok(report);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    // --- Campaign Export Endpoints ---

    [HttpGet("campaigns/csv")]
    public async Task<IActionResult> DownloadCampaignsCsv([FromQuery] string? period, [FromQuery] string? startDate, [FromQuery] string? endDate)
    {
        var email = GetUserEmail();
        var campaigns = await _campaignService.GetCampaignsAsync(email);
        var data = _exportService.ExportCampaignsToCsv(campaigns);
        return File(data, "text/csv", "campaigns.csv");
    }

    [HttpGet("campaigns/excel")]
    public async Task<IActionResult> DownloadCampaignsExcel([FromQuery] string? period, [FromQuery] string? startDate, [FromQuery] string? endDate)
    {
        var email = GetUserEmail();
        var campaigns = await _campaignService.GetCampaignsAsync(email);
        var data = _exportService.ExportCampaignsToExcel(campaigns);
        return File(data, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "campaigns.xlsx");
    }

    [HttpGet("campaigns/pdf")]
    public async Task<IActionResult> DownloadCampaignsPdf([FromQuery] string? period, [FromQuery] string? startDate, [FromQuery] string? endDate)
    {
        var email = GetUserEmail();
        var campaigns = await _campaignService.GetCampaignsAsync(email);
        var workspace = await _workspaceService.GetCurrentWorkspaceAsync(email);
        var data = _exportService.ExportCampaignsToPdf(campaigns, workspace.Name);
        return File(data, "application/pdf", "campaigns.pdf");
    }

    // --- Lead Export Endpoints ---
 
    [HttpGet("leads/csv")]
    public async Task<IActionResult> DownloadLeadsCsv([FromQuery] string? period, [FromQuery] string? startDate, [FromQuery] string? endDate, [FromQuery] long? userId)
    {
        var email = GetUserEmail();
        var leads = await _leadService.GetLeadsAsync(email, period, startDate, endDate, userId);
        var data = _exportService.ExportLeadsToCsv(leads);
        return File(data, "text/csv", "leads.csv");
    }

    [HttpGet("leads/excel")]
    public async Task<IActionResult> DownloadLeadsExcel([FromQuery] string? period, [FromQuery] string? startDate, [FromQuery] string? endDate, [FromQuery] long? userId)
    {
        var email = GetUserEmail();
        var leads = await _leadService.GetLeadsAsync(email, period, startDate, endDate, userId);
        var data = _exportService.ExportLeadsToExcel(leads);
        return File(data, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "leads.xlsx");
    }

    [HttpGet("leads/pdf")]
    public async Task<IActionResult> DownloadLeadsPdf([FromQuery] string? period, [FromQuery] string? startDate, [FromQuery] string? endDate, [FromQuery] long? userId)
    {
        var email = GetUserEmail();
        var leads = await _leadService.GetLeadsAsync(email, period, startDate, endDate, userId);
        var workspace = await _workspaceService.GetCurrentWorkspaceAsync(email);
        var data = _exportService.ExportLeadsToPdf(leads, workspace.Name);
        return File(data, "application/pdf", "leads.pdf");
    }

    [HttpGet("leads/{id}/pdf")]
    public IActionResult DownloadSingleLeadPdf(long id)
    {
        var data = _exportService.ExportSingleLeadPdf(id);
        return File(data, "application/pdf", $"lead_{id}_dossier.pdf");
    }

    private string GetUserEmail()
    {
        return User.FindFirstValue(ClaimTypes.Name) ?? User.FindFirstValue(ClaimTypes.Email) ?? string.Empty;
    }
}
